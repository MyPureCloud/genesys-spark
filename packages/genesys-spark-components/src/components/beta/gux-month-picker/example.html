<h1>gux-month-picker-beta</h1>

<div onchange="notify(event)" oninput="notify(event)">
  <h2>Default</h2>
  <gux-month-picker-beta value="2022-01"></gux-month-picker-beta>

  <h2>Disabled</h2>
  <gux-month-picker-beta value="2022-01" disabled></gux-month-picker-beta>

  <h2>Min & Max</h2>
  <gux-month-picker-beta
    value="2022-01"
    min="2021-03"
    max="2023-09"
  ></gux-month-picker-beta>

  <h2>Languages</h2>
  <div id="languages-container">
    <div id="sample"></div>
  </div>
</div>

<style
  onload="(function () {
    const languagesContainer = document.getElementById('languages-container');

    [
      'ar', 'cs', 'da', 'de', 'en', 'es-es', 'es', 'fi','fr-ca',
      'fr', 'he','it', 'ja', 'ko', 'nl', 'no', 'pl', 'pt-br',
      'pt-pt', 'ru', 'sv', 'th', 'tr', 'uk', 'zh-cn', 'zh-tw'
    ].forEach((lang) => {
      const html = window.toHTML([
        '<div class=language lang=' + lang + '>',
          '<h3>' + lang + '</h3>',
          '<div class=group>',
            '<div class=example>',
              '<h4>Value Set</h4>',
              '<gux-month-picker-beta value=2022-01></gux-month-picker-beta>',
            '</div>',
            '<div class=example>',
              '<h4>Blank</h4>',
              '<gux-month-picker-beta></gux-month-picker-beta>',
            '</div>',
          '</div>',
        '</div>'
      ].join(''));

      languagesContainer.appendChild(html);
    });
})()"
>
  #languages-container {
    display: flex;
    flex-direction: column;
  }

  .language {
    padding-block: 10px;
    padding-inline: 25px;
  }

  .group {
    display: flex;
  }

  .example {
    padding-inline-end: 10px;
  }
</style>
