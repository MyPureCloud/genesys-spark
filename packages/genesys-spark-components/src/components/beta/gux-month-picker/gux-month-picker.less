@import (reference) '../../../style/color-palette.less';
@import (reference) '../../../style/mixins.less';
@import (reference) '../../../style/focus.less';
@import (reference) '../../../style/spacing.less';

:host {
  display: inline-block;
}

.gux-target {
  display: inline-flex;
  padding: @gux-spacing-2xs @gux-spacing-xs @gux-spacing-2xs @gux-spacing-small;
  cursor: pointer;
  background-color: @gux-grey-90;
  .gux-form-field-border();

  &:focus-within {
    .gux-focus-ring();
  }

  .gux-display {
    display: inline-flex;
    min-width: 100px;
    padding-right: @gux-spacing-xs;
    color: @gux-black-40;

    .gux-spinner {
      outline: none;

      &:not(:first-child) {
        margin-left: @gux-spacing-2xs;
      }

      &:focus-visible {
        background: @gux-blue-100;
      }

      &.gux-empty {
        color: @gux-black-90;
      }
    }
  }

  .gux-popup-toggle {
    justify-content: right;
    padding: 2px;
    color: @gux-black-90;
    background: transparent;
    border: none;
    border-radius: 4px;

    &:focus-visible {
      .gux-focus-ring-small();
    }

    &:not(:disabled):focus,
    &:not(:disabled):hover {
      color: @gux-black-40;
      cursor: pointer;
    }

    gux-icon {
      width: 16px;
      height: 16px;
    }

    &.gux-expanded {
      color: @gux-blue-60;

      &:not(:disabled):focus-visible,
      &:not(:disabled):hover {
        color: @gux-blue-60;
      }
    }
  }
}
