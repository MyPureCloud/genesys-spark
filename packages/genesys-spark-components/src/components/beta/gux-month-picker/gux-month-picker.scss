:host {
  display: inline-block;
}

.gux-target {
  display: inline-flex;
  gap: var(--gse-ui-formControl-input-gap);
  padding: var(--gse-ui-formControl-textarea-padding);
  cursor: pointer;
  background-color: var(--gse-ui-calendarMenu-backgroundColor);
  border: var(--gse-ui-formControl-input-default-border-width)
    var(--gse-ui-formControl-input-default-border-style)
    var(--gse-ui-formControl-input-default-border-color);
  border-radius: var(--gse-ui-formControl-input-borderRadius);

  &:focus-within {
    border: var(--gse-ui-formControl-input-active-border-width)
      var(--gse-ui-formControl-input-active-border-style)
      var(--gse-ui-formControl-input-active-border-color);
    border-radius: var(--gse-ui-formControl-focusRing-borderRadius);
    outline: var(--gse-ui-formControl-input-focus-border-width)
      var(--gse-ui-formControl-input-focus-border-style)
      var(--gse-ui-formControl-input-focus-border-color);
    outline-offset: var(--gse-semantic-focusRing-offset);
  }

  .gux-display {
    display: inline-flex;
    min-width: 100px;
    padding-right: var(--gse-ui-formControl-input-gap);
    font-family: var(--gse-ui-formControl-label-text-fontFamily);
    font-size: var(--gse-ui-formControl-label-text-fontSize);
    font-weight: var(--gse-ui-formControl-label-text-fontWeight);
    line-height: var(--gse-ui-formControl-label-text-lineHeight);
    color: var(--gse-ui-formControl-label-labelColor);

    .gux-spinner {
      outline: none;

      &:not(:first-child) {
        margin-left: 4px;
      }

      &:focus-visible {
        background: var(--gse-ui-calendarMenu-month-hover-backgroundColor);
      }
    }
  }

  .gux-popup-toggle {
    justify-content: right;
    color: var(--gse-ui-monthPicker-calendarStates-defaultColor);
    background: transparent;
    border: none;

    &:focus-visible {
      border-radius: var(
        --gse-ui-calendarMenu-month-calendarButton-focusBorderRadius
      );
      outline: var(--gse-ui-monthPicker-calendarStates-focus-border-width)
        var(--gse-ui-monthPicker-calendarStates-focus-border-style)
        var(--gse-ui-monthPicker-calendarStates-focus-border-color);
      outline-offset: var(--gse-semantic-focusRing-offset);
    }

    &:not(:disabled):hover {
      color: var(--gse-ui-monthPicker-calendarStates-hoverColor);
      cursor: pointer;
    }

    &:not(:disabled):focus {
      color: var(--gse-ui-monthPicker-calendarStates-activeColor);
      cursor: pointer;
    }

    &:disabled {
      color: var(--gse-ui-monthPicker-calendarStates-disabledColor);
    }

    gux-icon {
      width: var(--gse-ui-icon-size-sm);
      height: var(--gse-ui-icon-size-sm);
    }

    &.gux-expanded {
      color: var(--gse-ui-monthPicker-calendarStates-activeColor);
    }
  }
}
