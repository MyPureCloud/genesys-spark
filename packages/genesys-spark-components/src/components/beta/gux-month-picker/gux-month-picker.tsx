/* eslint-disable @typescript-eslint/restrict-template-expressions */
import {
  Component,
  Element,
  h,
  JSX,
  Listen,
  Prop,
  State,
  Watch
} from '@stencil/core';

import { trackComponent } from '@utils/tracking/usage';
import { OnClickOutside } from '@utils/decorator/on-click-outside';
import {
  buildI18nForComponent,
  GetI18nValue,
  getDesiredLocale
} from '../../../i18n';
import { sparkIntl } from 'genesys-spark-utils';
import { useRegionalDates } from '../../../i18n/use-regional-dates';
import simulateNativeEvent from '@utils/dom/simulate-native-event';
import { afterNextRender } from '@utils/dom/after-next-render';
import {
  getCurrentISOYearMonth,
  getISOYearMonth,
  getYearMonthObject
} from '@utils/date/year-month-values';

import { GuxISOYearMonth } from '../../../utils/date/year-month-values';

import translationResources from './i18n/en.json';

const DateTimeFormats = { year: 'numeric', month: 'long' };

const YearFirstLocales = ['ar', 'he', 'ja', 'ko', 'zh-cn', 'zh-tw'];

@Component({
  styleUrl: 'gux-month-picker.scss',
  tag: 'gux-month-picker-beta',
  shadow: { delegatesFocus: false }
})
export class GuxMonthPicker {
  private i18n: GetI18nValue;
  private calendarToggleButtonElement: HTMLButtonElement;
  private monthCalendarElement: HTMLGuxMonthCalendarElement;
  private monthSpinnerElement: HTMLElement;
  private yearSpinnerElement: HTMLElement;

  @Element()
  private root: HTMLElement;

  /**
   * The current selected year and month in ISO8601 format (yyyy-mm)
   */
  @Prop({ mutable: true })
  value: GuxISOYearMonth;

  /**
   * The min year and month selectable in ISO8601 format (yyyy-mm)
   */
  @Prop()
  min: GuxISOYearMonth;

  /**
   * The max year and month selectable in ISO8601 format (yyyy-mm)
   */
  @Prop()
  max: GuxISOYearMonth;

  @Prop()
  disabled: boolean = false;

  @State()
  expanded: boolean = false;

  @State()
  locale: string;

  @Listen('keydown')
  onKeyDown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Escape':
        this.expanded = false;
        this.calendarToggleButtonElement.focus();
        break;
    }
  }

  @Watch('value')
  onValueUpdate(newValue: GuxISOYearMonth) {
    if (this.isOutOfBounds(newValue)) {
      if (this.isBeforeMin(newValue)) {
        this.value = this.min;
      } else {
        this.value = this.max;
      }
    } else {
      this.value = newValue;
    }

    simulateNativeEvent(this.root, 'input');
    simulateNativeEvent(this.root, 'change');
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  onClickOutside(): void {
    this.expanded = false;
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);
    if (useRegionalDates()) {
      this.locale = sparkIntl.determineDisplayLocale(this.root);
    } else {
      this.locale = getDesiredLocale(this.root);
    }
  }

  @Listen('focusout')
  onFocusout(event: FocusEvent): void {
    const focusIsOutsideComponent =
      event.relatedTarget && !this.root.contains(event.relatedTarget as Node);
    if (focusIsOutsideComponent) {
      this.expanded = false;
    }
  }

  private isOutOfBounds(value: GuxISOYearMonth): boolean {
    return this.isBeforeMin(value) || this.isAfterMax(value);
  }

  private isBeforeMin(value: GuxISOYearMonth): boolean {
    return this.min && this.min > value;
  }

  private isAfterMax(value: GuxISOYearMonth): boolean {
    return this.max && this.max < value;
  }

  private toggleCalendar(): void {
    this.expanded = !this.expanded;

    if (this.expanded) {
      afterNextRender(() => {
        void this.monthCalendarElement.guxFocus(this.value);
      });
    }
  }

  private onMonthCalendarInput(): void {
    this.value = this.monthCalendarElement.value;
    this.expanded = false;
    this.calendarToggleButtonElement.focus();
  }

  private incrementMonth(delta: 1 | -1): void {
    if (this.value) {
      const { year: currentYear, month: currentMonth } = getYearMonthObject(
        this.value
      );
      const newMonth = (((parseInt(currentMonth) + 11 + delta) % 12) + 1)
        .toString()
        .padStart(2, '0');

      this.value = getISOYearMonth(currentYear, newMonth);
    } else {
      this.value = getCurrentISOYearMonth();
    }
  }

  private incrementYear(delta: 1 | -1): void {
    if (this.value) {
      const { year: currentYear, month: currentMonth } = getYearMonthObject(
        this.value
      );
      const newYear = Math.max(Number(currentYear) + delta, 0).toString();

      this.value = getISOYearMonth(newYear, currentMonth);
    } else {
      this.value = getCurrentISOYearMonth();
    }
  }

  private onSpinnerKeyDown(
    event: KeyboardEvent,
    incrementor: (delta: 1 | -1) => void,
    nextFocusElement: HTMLElement
  ): void {
    switch (event.key) {
      case 'ArrowDown':
        event.preventDefault();
        incrementor(-1);
        break;
      case 'ArrowUp':
        event.preventDefault();
        incrementor(1);
        break;
      case 'ArrowLeft':
      case 'ArrowRight':
        event.preventDefault();
        nextFocusElement.focus();
        break;
      case 'Enter':
        event.preventDefault();
        this.expanded = true;
        afterNextRender(() => {
          void this.monthCalendarElement.guxFocus(this.value);
        });
        break;
      case ' ':
        event.preventDefault();
        break;
    }
  }

  private onSpinnerKeyUp(event: KeyboardEvent): void {
    event.stopPropagation();

    switch (event.key) {
      case ' ':
        event.preventDefault();
        this.expanded = true;
        afterNextRender(() => {
          void this.monthCalendarElement.guxFocus(this.value);
        });
        break;
    }
  }

  private onSpinnerClick(): void {
    this.expanded = true;
  }

  private getSpinnerLabel(period: 'month' | 'year'): string {
    if (this.value) {
      const { year, month } = getYearMonthObject(this.value);

      return new Date(Number(year), Number(month) - 1).toLocaleDateString(
        this.locale,
        { [period]: DateTimeFormats[period] }
      );
    }

    return this.i18n(period);
  }

  private getSpinnerValueNow(period: 'month' | 'year'): string {
    return this.value ? getYearMonthObject(this.value)[period] : '0';
  }

  private getSpinnerValueText(): string | boolean {
    if (this.value) {
      const { year, month } = getYearMonthObject(this.value);

      return new Date(Number(year), Number(month) - 1).toLocaleDateString(
        this.locale,
        { year: 'numeric', month: 'long' }
      );
    }

    return this.i18n('unset');
  }

  private renderMonthSpinnerButton(): JSX.Element {
    return (
      <div
        role="spinbutton"
        class="gux-spinner"
        tabIndex={this.disabled ? -1 : 0}
        onKeyDown={(e: KeyboardEvent) =>
          this.onSpinnerKeyDown(
            e,
            d => this.incrementMonth(d),
            this.yearSpinnerElement
          )
        }
        onKeyUp={(e: KeyboardEvent) => this.onSpinnerKeyUp(e)}
        onClick={() => this.onSpinnerClick()}
        ref={(el: HTMLElement) => (this.monthSpinnerElement = el)}
        aria-valuenow={this.getSpinnerValueNow('month')}
        aria-valuetext={this.getSpinnerValueText()}
        aria-valuemin="1"
        aria-valuemax="12"
        aria-label={this.i18n('month')}
      >
        {this.getSpinnerLabel('month')}
      </div>
    ) as JSX.Element;
  }

  private renderYearSpinnerButton(): JSX.Element {
    return (
      <div
        role="spinbutton"
        class="gux-spinner"
        tabIndex={this.disabled ? -1 : 0}
        onKeyDown={(e: KeyboardEvent) =>
          this.onSpinnerKeyDown(
            e,
            d => this.incrementYear(d),
            this.monthSpinnerElement
          )
        }
        onKeyUp={(e: KeyboardEvent) => this.onSpinnerKeyUp(e)}
        onClick={() => this.onSpinnerClick()}
        ref={(el: HTMLElement) => (this.yearSpinnerElement = el)}
        aria-valuenow={this.getSpinnerValueNow('year')}
        aria-valuetext={this.getSpinnerValueText()}
        aria-valuemin="0"
        aria-label={this.i18n('year')}
      >
        {this.getSpinnerLabel('year')}
      </div>
    ) as JSX.Element;
  }

  private renderSpinnerButtons(): JSX.Element {
    if (YearFirstLocales.includes(this.locale)) {
      return (
        <span class="gux-display">
          {this.renderYearSpinnerButton()}
          {this.renderMonthSpinnerButton()}
        </span>
      ) as JSX.Element;
    }

    return (
      <span class="gux-display">
        {this.renderMonthSpinnerButton()}
        {this.renderYearSpinnerButton()}
      </span>
    ) as JSX.Element;
  }

  private renderCalendarToggleButton(): JSX.Element {
    return (
      <button
        class={{
          'gux-popup-toggle': true,
          'gux-expanded': this.expanded
        }}
        ref={(el: HTMLButtonElement) => (this.calendarToggleButtonElement = el)}
        type="button"
        onClick={() => this.toggleCalendar()}
        disabled={this.disabled}
      >
        <gux-icon
          icon-name="fa/calendar-regular"
          decorative
          size="small"
        ></gux-icon>
        <gux-screen-reader-beta>
          {this.i18n('toggleCalendar')}
        </gux-screen-reader-beta>
      </button>
    ) as JSX.Element;
  }

  renderTarget(): JSX.Element {
    return (
      <div class="gux-target" slot="target">
        {this.renderSpinnerButtons()}
        {this.renderCalendarToggleButton()}
      </div>
    ) as JSX.Element;
  }

  private renderPopup(): JSX.Element {
    return (
      <gux-month-calendar
        slot="popup"
        ref={(el: HTMLGuxMonthCalendarElement) =>
          (this.monthCalendarElement = el)
        }
        onInput={() => this.onMonthCalendarInput()}
        value={this.value}
        min={this.min}
        max={this.max}
      />
    ) as JSX.Element;
  }

  render(): JSX.Element {
    return (
      <gux-popup
        expanded={this.expanded}
        disabled={this.disabled}
        exceed-target-width
      >
        {this.renderTarget()}
        {this.renderPopup()}
      </gux-popup>
    ) as JSX.Element;
  }
}
