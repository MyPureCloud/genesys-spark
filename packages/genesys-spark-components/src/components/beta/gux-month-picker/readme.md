# gux-month-picker



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                                     | Type                    | Default     |
| ---------- | ---------- | --------------------------------------------------------------- | ----------------------- | ----------- |
| `disabled` | `disabled` |                                                                 | `boolean`               | `false`     |
| `max`      | `max`      | The max year and month selectable in ISO8601 format (yyyy-mm)   | ``${string}-${string}`` | `undefined` |
| `min`      | `min`      | The min year and month selectable in ISO8601 format (yyyy-mm)   | ``${string}-${string}`` | `undefined` |
| `value`    | `value`    | The current selected year and month in ISO8601 format (yyyy-mm) | ``${string}-${string}`` | `undefined` |


## Dependencies

### Depends on

- [gux-icon](../../stable/gux-icon)
- [gux-month-calendar](gux-month-calendar)
- [gux-popup](../../stable/gux-popup)

### Graph
```mermaid
graph TD;
  gux-month-picker-beta --> gux-icon
  gux-month-picker-beta --> gux-month-calendar
  gux-month-picker-beta --> gux-popup
  gux-month-calendar --> gux-icon
  gux-month-calendar --> gux-month-list-item
  gux-month-calendar --> gux-month-list
  style gux-month-picker-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
