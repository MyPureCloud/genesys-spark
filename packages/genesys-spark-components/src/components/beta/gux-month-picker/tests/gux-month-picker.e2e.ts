import { newSparkE2EPage, a11yCheck } from '../../../../test/e2eTestUtils';
import { renderConfigs } from './gux-month-picker.common';

const axeExclusions = [];

describe('gux-month-picker-beta', () => {
  describe('#render', () => {
    renderConfigs.forEach(({ description, html }) => {
      it(description, async () => {
        const page = await newSparkE2EPage({ html });
        const element = await page.find('gux-month-picker-beta');

        await a11yCheck(page, axeExclusions);

        expect(element.outerHTML).toMatchSnapshot();
      });
    });
  });

  describe('gux-month-calendar year navigation', () => {
    it('should update year label when clicking year change buttons', async () => {
      const page = await newSparkE2EPage({
        html: '<gux-month-calendar value="2023-06"></gux-month-calendar>'
      });

      await page.waitForSelector('gux-month-calendar');

      const getYearLabel = () =>
        page.evaluate(() => {
          const calendar = document.querySelector('gux-month-calendar');
          return calendar.shadowRoot.querySelector('[data-testid="year-label"]')
            .textContent;
        });

      const clickYearButton = async (index: number) => {
        await page.evaluate(buttonIndex => {
          const calendar = document.querySelector('gux-month-calendar');
          const buttons =
            calendar.shadowRoot.querySelectorAll('.gux-year-change');
          (buttons[buttonIndex] as HTMLButtonElement).click();
        }, index);
        await page.waitForChanges();
      };

      expect(await getYearLabel()).toContain('2023');

      await clickYearButton(1); // next year
      expect(await getYearLabel()).toContain('2024');

      await clickYearButton(0); // previous year
      await clickYearButton(0); // previous year again
      expect(await getYearLabel()).toContain('2022');
    });
  });
});
