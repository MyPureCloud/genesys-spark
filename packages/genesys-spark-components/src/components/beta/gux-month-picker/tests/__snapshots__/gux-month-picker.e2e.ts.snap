// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-month-picker-beta #render should render a disabled gux-month-picker-beta 1`] = `"<gux-month-picker-beta disabled="" value="2025-01" hydrated=""><template shadowrootmode="open"><gux-popup exceed-target-width="" hydrated=""><div class="gux-target" slot="target"><span class="gux-display"><div role="spinbutton" class="gux-spinner" tabindex="-1" aria-valuenow="01" aria-valuetext="January 2025" aria-valuemin="1" aria-valuemax="12" aria-label="Month">January</div><div role="spinbutton" class="gux-spinner" tabindex="-1" aria-valuenow="2025" aria-valuetext="January 2025" aria-valuemin="0" aria-label="Year">2025</div></span><button class="gux-popup-toggle" type="button" disabled=""><gux-icon icon-name="fa/calendar-regular" screenreader-text="Toggle month calendar view" size="small" hydrated=""></gux-icon></button></div><gux-month-calendar slot="popup" hydrated=""></gux-month-calendar></gux-popup></template></gux-month-picker-beta>"`;

exports[`gux-month-picker-beta #render should render gux-month-picker-beta 1`] = `"<gux-month-picker-beta value="2025-01" hydrated=""><template shadowrootmode="open"><gux-popup exceed-target-width="" hydrated=""><div class="gux-target" slot="target"><span class="gux-display"><div role="spinbutton" class="gux-spinner" tabindex="0" aria-valuenow="01" aria-valuetext="January 2025" aria-valuemin="1" aria-valuemax="12" aria-label="Month">January</div><div role="spinbutton" class="gux-spinner" tabindex="0" aria-valuenow="2025" aria-valuetext="January 2025" aria-valuemin="0" aria-label="Year">2025</div></span><button class="gux-popup-toggle" type="button"><gux-icon icon-name="fa/calendar-regular" screenreader-text="Toggle month calendar view" size="small" hydrated=""></gux-icon></button></div><gux-month-calendar slot="popup" hydrated=""></gux-month-calendar></gux-popup></template></gux-month-picker-beta>"`;
